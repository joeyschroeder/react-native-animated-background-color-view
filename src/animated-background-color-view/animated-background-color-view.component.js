import { Animated, Easing, ViewPropTypes } from 'react-native';
import React, { Component } from 'react';

import Color from 'color';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';

export class AnimatedBackgroundColorView extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
    color: PropTypes.string,
    delay: PropTypes.number,
    duration: PropTypes.number,
    easing: PropTypes.func,
    initialColor: PropTypes.string,
    style: ViewPropTypes.style,
  };

  static defaultProps = {
    children: null,
    color: '#00aced',
    delay: 0,
    duration: 2400,
    easing: Easing.out(Easing.exp),
    initialColor: null,
    style: null,
  };

  constructor(props) {
    super(props);
    const { initialColor, color } = props;

    const currentColor = initialColor || color;

    this.animation = new Animated.Value(0);
    this.currentColor = currentColor;
  }

  componentDidMount() {
    this.animate();
  }

  shouldComponentUpdate(nextProps) {
    const { children, color } = this.props;
    return !isEqual([color, children], [nextProps.color, nextProps.children]);
  }

  componentDidUpdate() {
    this.animate();
  }

  getBackgroundColor() {
    const { color } = this.props;

    const backgroundColor = this.animation.interpolate({
      inputRange: [0, 1],
      outputRange: [this.currentColor, color],
    });

    return backgroundColor;
  }

  getCurrentColor(animationFinished = true, mixValue = 1) {
    const { color } = this.props;

    if (animationFinished) return color;
    return this.getMixedColorValue(mixValue);
  }

  getMixedColorValue(mixValue = 1) {
    const { color } = this.props;

    const nextColor = Color(color);
    const currentColor = Color(this.currentColor);

    const mixedColor = currentColor.mix(nextColor, mixValue).hex();

    return mixedColor;
  }

  shouldAnimate() {
    const { color } = this.props;
    return color !== this.currentColor;
  }

  stopAnimationAndReturnAnimationValue() {
    this.animation.stopAnimation((value) => {
      return value;
    });
  }

  animate() {
    if (!this.shouldAnimate()) return;

    const { delay, duration, easing } = this.props;

    const stopValue = this.stopAnimationAndReturnAnimationValue();
    this.animation.setValue(0);

    Animated.timing(this.animation, {
      delay,
      duration,
      easing,
      toValue: 1,
      useNativeDriver: false,
    }).start(({ finished }) => {
      this.currentColor = this.getCurrentColor(finished, stopValue);
    });
  }

  render() {
    const { children, color, delay, duration, easing, initialColor, style, ...other } = this.props;

    const backgroundColor = this.getBackgroundColor();
    const viewStyle = [style, { backgroundColor }];

    return (
      <Animated.View {...other} style={viewStyle}>
        {children}
      </Animated.View>
    );
  }
}
