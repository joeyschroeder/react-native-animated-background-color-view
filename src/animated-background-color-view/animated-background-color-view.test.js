import Color from 'color';
import React from 'react';
import { View } from 'react-native';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import { AnimatedBackgroundColorView } from './animated-background-color-view.component';

describe('<AnimatedBackgroundColorView />', () => {
  describe('this.componentDidMount', () => {
    const component = shallow(<AnimatedBackgroundColorView />);
    const animate = jest.fn();

    component.instance().animate = animate;
    component.instance().componentDidMount();

    it('should call this.animate', () => {
      expect(animate).toHaveBeenCalled();
    });
  });

  describe('this.shouldComponentUpdate', () => {
    it('should return true when this.props.children or this.props.color have changed', () => {
      const nextProps = { children: <View />, color: 'test-1' };
      const component = shallow(
        <AnimatedBackgroundColorView color="test-2">{nextProps.children}</AnimatedBackgroundColorView>
      );

      const result = component.instance().shouldComponentUpdate(nextProps);
      expect(result).toEqual(true);
    });

    it('should return false when this.props.children or this.props.color have not changed', () => {
      const nextProps = { children: <View />, color: 'test-1' };
      const component = shallow(
        <AnimatedBackgroundColorView color={nextProps.color}>{nextProps.children}</AnimatedBackgroundColorView>
      );

      const result = component.instance().shouldComponentUpdate(nextProps);
      expect(result).toEqual(false);
    });
  });

  describe('this.componentDidUpdate', () => {
    const component = shallow(<AnimatedBackgroundColorView />);
    const animate = jest.fn();

    component.instance().animate = animate;
    component.instance().componentDidUpdate();

    it('should call this.animate', () => {
      expect(animate).toHaveBeenCalled();
    });
  });

  describe('this.getCurrentColor', () => {
    it('should return this.props.color when animationFinished param is true', () => {
      const color = '#000';
      const component = shallow(<AnimatedBackgroundColorView color={color} />);
      const result = component.instance().getCurrentColor(true);

      expect(result).toEqual(color);
    });

    it('should return mixed color when animationFinished param is false', () => {
      const color = '#000';
      const currentColor = '#fff';
      const mixValue = 0.5;
      const component = shallow(<AnimatedBackgroundColorView color={color} />);
      component.instance().currentColor = currentColor;
      const result = component.instance().getCurrentColor(false, mixValue);

      expect(result).toEqual(Color(currentColor).mix(Color(color), mixValue).hex());
    });
  });

  it('should render correctly', () => {
    const component = create(<AnimatedBackgroundColorView />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
