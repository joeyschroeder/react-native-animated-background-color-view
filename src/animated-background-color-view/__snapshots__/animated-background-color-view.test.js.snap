// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimatedBackgroundColorView /> should render correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "backgroundColor": "rgba(0, 172, 237, 1)",
    }
  }
/>
`;
